db/kv_checksum.h

```diff
@@ -0,0 +1,70 @@
+//  Copyright (c) 2020-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include "db/dbformat.h"
+#include "rocksdb/types.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// Holds checksums of user keys/values and copies of internal metadata for
+// protecting integrity of data as it is processed.
+class KvProtectionInfo {
```

Comment:
This class is very complicated. Is it possible to start with something simpler. E.g., we hard code how the checksum is generated and generated to a single checksum, rather than several checksums and each for a field?

Reply:
Yes, one option was to have a checksum that covers all the key/value/timestamp buffers and all the metadata. On the design doc, Peter pointed out that would catch cases like key and value being swapped. However, it is a bit difficult to work with as the components that need to be covered change during processing. For example, we add seqno during the `WriteBatch` -> `MemTable` processing. I think we also lose CF ID at some point.

Also we'd need to add back the xxh3 streaming APIs in order to implement that efficiently. This problem is fixable though.

Reply:
How about regenerate the checksum every time we add something? Or extending it?

Reply:
I read Peter's comment (https://github.com/facebook/rocksdb/pull/7748#discussion_r541215097) after writing my reply here. I think that design provides a good way to embed the covered checksum components into the type system, without having to do extra checksum computation (as with regenerating) or worrying about the order the fields were added (as with extending).